arch/risc-v/src/common/riscv_mtimer.c

```diff
@@ -0,0 +1,306 @@
+ *
+ ****************************************************************************/
+
+static int riscv_mtimer_cancel(struct oneshot_lowerhalf_s *lower,
+                               struct timespec *ts)
+{
+  struct riscv_mtimer_lowerhalf_s *priv =
+    (struct riscv_mtimer_lowerhalf_s *)lower;
+  uint64_t mtime;
+
+  riscv_mtimer_set_mtimecmp(priv, UINT64_MAX);
+
+  mtime = riscv_mtimer_get_mtime(priv);
+  if (priv->alarm > mtime)
+    {
+      uint64_t nsec = (priv->alarm - mtime) *
+                      NSEC_PER_SEC / priv->freq;
+
+      ts->tv_sec  = nsec / NSEC_PER_SEC;
+      ts->tv_nsec = nsec % NSEC_PER_SEC;
+    }
+  else
+    {
+      ts->tv_sec  = 0;
+      ts->tv_nsec = 0;
+    }
+
+  priv->alarm    = 0;
+  priv->callback = NULL;
+  priv->arg      = NULL;
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: riscv_mtimer_current
+ *
+ * Description:
+ *  Get the current time.
+ *
+ * Input Parameters:
+ *   lower   Caller allocated instance of the oneshot state structure.  This
+ *           structure must have been previously initialized via a call to
+ *           oneshot_initialize();
+ *   ts      The location in which to return the current time. A time of zero
+ *           is returned for the initialization moment.
+ *
+ * Returned Value:
+ *   Zero (OK) is returned on success, a negated errno value is returned on
+ *   any failure.
+ *
+ ****************************************************************************/
+
+static int riscv_mtimer_current(struct oneshot_lowerhalf_s *lower,
+                                struct timespec *ts)
+{
+  struct riscv_mtimer_lowerhalf_s *priv =
+    (struct riscv_mtimer_lowerhalf_s *)lower;
+  uint64_t mtime = riscv_mtimer_get_mtime(priv);
+  uint64_t nsec = mtime * NSEC_PER_SEC / priv->freq;
+
+  ts->tv_sec  = nsec / NSEC_PER_SEC;
+  ts->tv_nsec = nsec % NSEC_PER_SEC;
+
+  return 0;
+}
+
+static int riscv_mtimer_interrupt(int irq, void *context, void *arg)
+{
+  struct riscv_mtimer_lowerhalf_s *priv = arg;
+
+  riscv_mtimer_set_mtimecmp(priv, UINT64_MAX);
+  if (priv->callback != NULL)
+    {
+      priv->callback(&priv->lower, priv->arg);
+    }
+
+  return 0;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+struct oneshot_lowerhalf_s *
+riscv_mtimer_initialize(uintptr_t mtime, uintptr_t mtimecmp,
+                        int irq, uint64_t freq)
+{
+  struct riscv_mtimer_lowerhalf_s *priv;
+
+  priv = kmm_zalloc(sizeof(*priv));
+  if (priv != NULL)
+    {
+      priv->lower.ops = &g_riscv_mtimer_ops;
+      priv->mtime     = mtime;
+      priv->mtimecmp  = mtimecmp;
+      priv->freq      = freq;
+
+      riscv_mtimer_set_mtimecmp(priv, UINT64_MAX);
+      irq_attach(irq, riscv_mtimer_interrupt, priv);
+      up_enable_irq(irq);
+    }
+
+  return &priv->lower;
```

Comment:
So i case if `kmm_zalloc` we will de-reference NULL pointer?

Reply:
> So i case if `kmm_zalloc` we will de-reference NULL pointer?

There isn't de-reference(*) operation. But, let me change to the cast which is more clear.

Reply:
Done.

Reply:
Sorry, I missed that `lower` is the first element in the struct. Indeed compiler will eliminate de-reference operation in such case. But anyway now it is more clear without need to go into details. Thank you